File: binarytree/0226-invert.cpp
/*
 * leetcode 226 invert
 *
 * binary tree template
 *
 */

class Solution {
public:
    // the same idea as maxDepth problem
    // recursive
    // think about the current level only
    void invert(TreeNode* root) {
        // base condition or stop condition
        // if null, no need to invert
        if (!root) {
            return;
        }
        // 
        invert(root->left);
        invert(root->right);
        
        // 
        TreeNode* left = root->left;
        TreeNode* right = root->right;
        root->left = right;
        root->right = left;
    }

    TreeNode* invertTree(TreeNode* root) {
        invert(root);
        return root;
    }
};

/*

https://leetcode.cn/problems/maximum-depth-of-binary-tree/solutions/1413111/yi-tao-mo-ban-jie-jue-duo-ge-by-zhang-xi-yqep/

*/

File: binarytree/0226-invert.py
"""
leetcode 226
invert a binary tree

pre-order: root, left, right

1 exchange left and right subtrees, 
2 left-subtree, 
3 right-subtree

"""

#
# recusive
#
class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        if root == None:
            return None
        root.left, root.right = root.right, root.left
        self.invertTree(root.left)
        self.invertTree(root.right)
        return root

#
# iterative
#
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root == None:
            return None
        
        stack = [root]
        while stack:
            node = stack.pop()
            node.left, node.right = node.right, node.left
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        return root
    
